### DISCUSSION TITLE
Logitech Buttons

### DISCUSSION THREAD

**Discussion by azwillnj** [hrvach/deskhop#166] [General]
There are some mentions to Logitech special keys, volume, play/pause, ect. in these discussions. It looks like a build called deskhop-test2.zip has it working from this message: https://github.com/hrvach/deskhop/discussions/94#discussioncomment-9861611. It also looks like there's a fork called DeskHopL that applies these changes with some other things I don't want. 

I am looking for the latest standard deskhop firmware that supports the Logitech special keys. Can you please point me in the direction of that? I am currently running 0.65 the special keys are not working.

**Comment by hrvach** [comment]
Can you please share what Logitech device are you trying it with? I will borrow one Logitech keyboard and re-test. Last time I tried it the special keys were working, but I'm known to occasionally break things (not on purpose) :)

**Reply by azwillnj** [reply]
Logitech MX Keys keyboard with a unifying dongle and a Logitech MX Master 3S mouse with a Bolt dongle. Yes, there are two separate dongles and they are not compatible with each other. Both the mouse and the keyboard have special buttons and functions and none are working. I can provide any other information you need or dumps just let me know what you want and how I can capture it.

Other than the special buttons this thing is amazing and is exactly what I was looking for. Thanks!

**Comment by alankeny** [comment]
@azwillnj I reported a problem with Logitech Media keys on [issue #150](https://github.com/hrvach/deskhop/issues/150).  The last version that worked for me was [release 0.63](https://github.com/hrvach/deskhop/releases/tag/v0.63).  Mute, volume up, volume down, play/pause etc all work for me with a Logitech MX keys keyboard on that release.

If you're already running a later release, you may run into the same problem I did while trying to downgrade to 0.63.  The two sides of the Deskhop will upgrade each other.  If you downgrade to 0.63 and the other side is running something later, it will immediately upgrade the 0.63 side to the later release.  I had to make sure only one side was connected at a time and use the older firmware upgrade method with the button.  That got everything back to 0.63 and working media keys.

**Reply by azwillnj** [reply]
@hrvach I know this is asking a lot, but can you do a new release with the logitech fix? Any testing I can do to help?

Also just saw you are from Munich! I'm very jealous, I studied there, and actually got married there (in Augustiner Keller!). Have an Augustiner for me!

**Reply by hrvach** [reply]
It's no problem, I was just waiting to get some Logitech gear, should be here tomorrow so I can get on with the bugfix. I kindly ask for a bit more time and thanks for your patience. One Augustiner coming up! :)

**Comment by hrvach** [comment]
Can you guys try it now? [deskhop.zip](https://github.com/user-attachments/files/17592881/deskhop.zip)
I think the keys should start working again, broke it in a dumb way. Please test drive this one and report back, then we'll push the fix officially.

@alankeny @azwillnj

**Reply by azwillnj** [reply]
It works! Thanks so much for the fix! All special buttons are working, volume, tracks, even the dumb one that launches the calculator.

**Comment by hrvach** [comment]
Please check mouse, stability, etc. Test drive it a bit and we'll update it. I haven't noticed it because my keyboard behaves differently and didn't set protocol to mouse for packets with report_id set. Thank you for your patience and hopefully there would be more time soon to improve the firmware some more.

**Reply by alankeny** [reply]
I can confirm it's working here too.  I tested with a Logitech MX keys keyboard with M525 mouse and a K400 keyboard / touchpad combination.  Everything is working as expected.

Thanks so much!

**Reply by azwillnj** [reply]
The only thing I can see that's not working is the Logi Options+ app, it's the Logitech app that allows you to set preferences and special features of the devices. I don't think it's possible for that to work with deskhop though and if I need to change something I can just remove it from deskhop and plug it into the computer directly. Thanks so much for the fw update!

**Reply by alankeny** [reply]
@azwillnj from what I can see using `lsusb` on Linux, the computer doesn't see the Logitech devices when they're connected through the Deskhop.  The computer sees a different USB device that the Deskhop provides, so it can do the device switching.

If you plug the Logitech devices directly into the computer, you should be able to change and save any preferences that are stored directly on the devices.  For example I can pair and unpair mice and keyboards with unifying receivers when the receiver directly connected to the computer, and they stay connected when the receiver is connected to the Deskhop.  Any features that require the Logitech app to be running probably won't continue to work when the devices are connected through the Deskhop, because the app won't recognize them as Logitech devices any more.

**Reply by ronalith** [reply]
New user here, but I can confirm the Logitech MX Master 2S receives the thumb button oddly in Mac OS. Seems to be working okay in Windows, behaving like an Alt+Tab, which is the default (no software installed) when directly plugged in. I'm unsure what the default behaviour is in Mac OS when no software installed.

**Reply by hrvach** [reply]
> New user here, but I can confirm the Logitech MX Master 2S receives the thumb button oddly in Mac OS. Seems to be working okay in Windows, behaving like an Alt+Tab, which is the default (no software installed) when directly plugged in. I'm unsure what the default behaviour is in Mac OS when no software installed.

Try the latest firmware, does that change the button behavior in any way?

**Reply by ronalith** [reply]
That's the one I'm running. version 0.66

**Comment by azwillnj** [comment]
I've been using this since release and have noticed no strange behavior, seems to be working perfectly, thanks again.

Thought of another feature request. Could you have all the media commands always forward to one of the computers? I am using this on my personal and work computers and the personal computer is where the sound system is connected and the only media is played. If you could select a computer in the web configurator that would always receive the media commands that would be very handy.

**Reply by hrvach** [reply]
That is a good suggestion indeed, I've noticed muting the wrong computer too a few times haha... can you please be so kind to put it https://github.com/hrvach/deskhop/discussions/27 here and I'll hopefully get some time off to work on the software side of things soon. Very happy to hear it works well for you!

**Comment by jalmeroth** [comment]
Everyone seems to be happy now and I can also confirm that my Logitech MX Keyboard (incl. Media Keys) and my Logitech MX Master 3S (incl. horizontal scrolling) are both working fine. Can we close this?

**Reply by ronalith** [reply]
MX Master 2S user here, in both Windows & Mac. Horizontal scrolling isn't working for me in neither system, and, thumb button (exposé) doesn't behave as default would on Mac OS, though I changed the shortcut in Mac OS to match what the mouse sends. Using v0.66.

**Reply by hrvach** [reply]
v0.66 still doesn't have the fix for horizontal scroll, have one PR with fix for relative mode to integrate before releasing a new version.

**Reply by hrvach** [reply]
@ronalith please try https://github.com/hrvach/deskhop/releases/tag/v0.67 and be so kind to report back. Thanks!
